protostar/commands/test/starkware/cheatable_execute_entry_point.py

```diff
@@ -0,0 +1,197 @@
+        )
+
+        # Run the specified contract entry point with given calldata.
+        with wrap_with_stark_exception(code=StarknetErrorCode.SECURITY_ERROR):
+            runner = CairoFunctionRunner(program=contract_class.program, layout="all")
+        os_context = os_utils.prepare_os_context(runner=runner)
+
+        # Extract pre-fetched contract state from carried state.
+        pre_run_contract_carried_state = state.contract_states[self.contract_address]
+        contract_state = pre_run_contract_carried_state.state
+        contract_state.assert_initialized(contract_address=self.contract_address)
+
+        starknet_storage = BusinessLogicStarknetStorage(
+            commitment_tree=contract_state.storage_commitment_tree,
+            ffc=state.ffc,
+            # Pass a copy of the carried storage updates (instead of a reference) - note that
+            # pending_modifications may be modified during the run as a result of an internal call.
+            pending_modifications=dict(pre_run_contract_carried_state.storage_updates),
+            loop=loop,
+        )
+
+        initial_syscall_ptr = cast(
+            RelocatableValue, os_context[starknet_abi.SYSCALL_PTR_OFFSET]
+        )
+
+        # --- MODIFICATIONS START --- # TODO
+        syscall_handler = CheatableSysCallHandler(
+            execute_entry_point_cls=CheatableExecuteEntryPoint,
+            tx_execution_context=tx_execution_context,
+            state=state,
+            caller_address=self.caller_address,
+            contract_address=self.contract_address,
+            starknet_storage=starknet_storage,
+            general_config=general_config,
+            initial_syscall_ptr=initial_syscall_ptr,
+        )
+
+        deployment_manager = DeploymentManager(
+            execute_entry_point_cls=CheatableExecuteEntryPoint,
+            tx_execution_context=tx_execution_context,
+            state=state,
+            caller_address=self.caller_address,
+            contract_address=self.contract_address,
+            starknet_storage=starknet_storage,
+            general_config=general_config,
+            initial_syscall_ptr=initial_syscall_ptr,
+        )
+
+        # Positional arguments are passed to *args in the 'run_from_entrypoint' function.
+        entry_points_args = [
+            self.entry_point_selector,
+            os_context,
+            len(self.calldata),
+            self.calldata,
+        ]
+
+        try:
+            runner.run_from_entrypoint(
+                entry_point.offset,
+                *entry_points_args,
+                hint_locals={
+                    "__storage": starknet_storage,
+                    "syscall_handler": syscall_handler,
+                    "declare": deployment_manager.declare,
+                    "prepare": deployment_manager.prepare_declared,
+                    "deploy": deployment_manager.deploy_prepared,
+                    "deploy_contract": build_deploy_contract(deployment_manager),
```

Comment:
As we talked I don't agree with you, cheatcodes modify the carried state not syscall handler, it would break any encapsulation of syscall handler. But we can discuss it further tommorow...

Reply:
I don't see how it is related to the comment above.

Anyway, how do you plan to reimplement e.g. `warp`, `roll`, `mock_call` cheatcodes? You need to have one syscall handler. How do you plan to merge cheatcodes into one cheatable syscall handler?

Reply:
>  You need to have one syscall handler.
I don't understand why. I want do something similar to CQRS here, this way objects will have small responsibilities. I don't want to do everything at once in a single PR. We probably will have to split DeploymentManager in to three separate classes. `DeclareCheatcode`, `PrepareCheatcode`, `DeployCheatcode`. Lot's of stuff in this PR is unpolished as I did this as an experiment.

Reply:
> I don't understand why. 

You need to have syscall handler to react on some events. For example, we need to replace syscall handler to modify behavior of `get_block_timestamp` and `get_block_number`. If `warp` and `roll` cheatcodes extend sys call handler we have to choose which one should replace the original sys call handler. 
 
```cairo
func get_block_timestamp{syscall_ptr : felt*}() -> (block_timestamp : felt):
    let syscall = [cast(syscall_ptr, GetBlockTimestamp*)]
    assert syscall.request = GetBlockTimestampRequest(selector=GET_BLOCK_TIMESTAMP_SELECTOR)
    %{ syscall_handler.get_block_timestamp(segments=segments, syscall_ptr=ids.syscall_ptr) %}
    let syscall_ptr = syscall_ptr + GetBlockTimestamp.SIZE
    return (block_timestamp=syscall.response.block_timestamp)
end

func get_block_number{syscall_ptr : felt*}() -> (block_number : felt):
    let syscall = [cast(syscall_ptr, GetBlockNumber*)]
    assert syscall.request = GetBlockNumberRequest(selector=GET_BLOCK_NUMBER_SELECTOR)
    %{ syscall_handler.get_block_number(segments=segments, syscall_ptr=ids.syscall_ptr) %}
    let syscall_ptr = syscall_ptr + GetBlockNumber.SIZE
    return (block_number=syscall.response.block_number)
end
```

>  I don't want to do everything at once in a single PR. 

I'll approve this PR just to deliver features in the next week. We shouldn't mix features and refactoring/experiments in the same PR.

> We probably will have to split DeploymentManager in to three separate classes. DeclareCheatcode, PrepareCheatcode, DeployCheatcode.

+1

Reply:
> You need to have syscall handler to react on some events. For example, we need to replace syscall handler to modify behavior of get_block_timestamp and get_block_number. If warp and roll cheatcodes extend sys call handler we have to choose which one should replace the original sys call handler.

original syscallhanlder will be replaced by CheatedSyscallHandler, cheatcodes are repsonsible for modifing state not recating to modifications.